File: Session_02_Recurrences/multi_language_implementations/recurrence_analysis.js
const readline = require('readline');

function printIndent(depth) {
    process.stdout.write("  |  ".repeat(depth));
}

function getSimulatedAddress(depth) {
    const baseAddress = 0x7ffd5eb3a000;
    const offset = depth * 0x30;
    return "0x" + (baseAddress - offset).toString(16);
}

// Factorial function
function calculateFactorial(n, depth) {
    printIndent(depth);
    const address = getSimulatedAddress(depth);
    console.log(`Entering factorial(${n}). Address of n: ${address}`);
    if (n <= 1) {
        printIndent(depth);
        console.log(`Base Case reached for factorial(${n}). Returning 1.`);
        return 1n;
    }
    const subResult = calculateFactorial(n - 1, depth + 1);
    const result = BigInt(n) * subResult;
    printIndent(depth);
    console.log(`Exiting factorial(${n}) -> returning ${n} * ${subResult} = ${result}`);
    return result;
}

// Fibonacci function
function calculateFibonacci(n, depth, branch) {
    printIndent(depth);
    const address = getSimulatedAddress(depth + 10);
    console.log(`Entering fibonacci(${n}) [${branch} branch]. Address of n: ${address}`);
    if (n === 0) {
        printIndent(depth);
        console.log("Base Case: fibonacci(0) = 0. Returning.");
        return 0;
    }
    if (n === 1) {
        printIndent(depth);
        console.log("Base Case: fibonacci(1) = 1. Returning.");
        return 1;
    }
    const left = calculateFibonacci(n - 1, depth + 1, "Left");
    const right = calculateFibonacci(n - 2, depth + 1, "Right");
    const result = left + right;
    printIndent(depth);
    console.log(`Exiting fibonacci(${n}) -> returning ${left} + ${right} = ${result}`);
    return result;
}

function main() {
    const val = 5;
    console.log("=== Recursion & Stack Frame Visualizer ===\n");
    console.log(`--- PART 1: Linear Recursion - Factorial(${val}) ---`);
    const fact = calculateFactorial(val, 0);
    console.log(`\nFactorial of ${val} is: ${fact}\n`);
    console.log("--------------------------------------------------\n");
    console.log("--- PART 2: Tree Recursion - Fibonacci(4) ---");
    const fib = calculateFibonacci(4, 0, "Root");
    console.log(`\nFibonacci of 4 is: ${fib}\n`);
    
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    rl.question("Press Enter to exit...", () => {
        rl.close();
    });
}

main();

File: Session_02_Recurrences/multi_language_implementations/recurrence_analysis.rs
use std::io::{self, Write};

fn print_indent(depth: i32) {
    for _ in 0..depth {
        print!("  |  ");
    }
}

fn calculate_factorial(n: i32, depth: i32) -> i64 {
    print_indent(depth);
    println!("Entering factorial({}). Address of n: {:p}", n, &n);
    if n <= 1 {
        print_indent(depth);
        println!("Base Case reached for factorial({}). Returning 1.", n);
        return 1;
    }
    let sub_result = calculate_factorial(n - 1, depth + 1);
    let result = n as i64 * sub_result;
    print_indent(depth);
    println!("Exiting factorial({}) -> returning {} * {} = {}", n, n, sub_result, result);
    return result;
}

fn calculate_fibonacci(n: i32, depth: i32, branch: &str) -> i32 {
    print_indent(depth);
    println!("Entering fibonacci({}) [{} branch]. Address of n: {:p}", n, branch, &n);
    if n == 0 {
        print_indent(depth);
        println!("Base Case: fibonacci(0) = 0. Returning.");
        return 0;
    }
    if n == 1 {
        print_indent(depth);
        println!("Base Case: fibonacci(1) = 1. Returning.");
        return 1;
    }
    let left = calculate_fibonacci(n - 1, depth + 1, "Left");
    let right = calculate_fibonacci(n - 2, depth + 1, "Right");
    let result = left + right;
    print_indent(depth);
    println!("Exiting fibonacci({}) -> returning {} + {} = {}", n, left, right, result);
    return result;
}

fn main() {
    let val = 5;
    println!("=== Recursion & Stack Frame Visualizer ===\n");
    println!("--- PART 1: Linear Recursion - Factorial({}) ---", val);
    let fact = calculate_factorial(val, 0);
    println!("\nFactorial of {} is: {}\n", val, fact);
    println!("--------------------------------------------------\n");
    println!("--- PART 2: Tree Recursion - Fibonacci(4) ---");
    let fib = calculate_fibonacci(4, 0, "Root");
    println!("\nFibonacci of 4 is: {}\n", fib);
    
    print!("Press Enter to exit...");
    let _ = io::stdout().flush();
    let mut input = String::new();
    let _ = io::stdin().read_line(&mut input);
}
